royalur/lut/reader.py

```diff
@@ -35,56 +43,102 @@ def get_metadata(self) -> Dict[str, Any]:
+
+        :param key: The key to look up.
+        :return: The value associated with the key.
+        """
+        index, _, _ = self._find_key_index(map_index, key)
+        if index is None:
+            raise KeyError(f"Key {key} not found in look-up table")
+        return self._get_value_at_index(map_index, index)["value"]
+
+    def lookup_detailed(self, map_index: int, key: int):
         """
         Look up a value in the look-up table.
 
         :param key: The key to look up.
         :return: The value associated with the key.
         """
-        index = self._find_index(key)
+        start_time = time.time()
+        index, key_lookup, steps = self._find_key_index(map_index, key)
         if index is None:
             raise KeyError(f"Key {key} not found in look-up table")
-        value_size = self._value_size
-        return int.from_bytes(
-            self._values[index * value_size: (index + 1) * value_size],
-            byteorder="big",
-            signed=True,
+        lookup = self._get_value_at_index(map_index, index)
+        return {
+            "index_of_key": index,
+            "key_binary_search_steps": steps,
+            "value_lookup_details": lookup,
+            "key_lookup_details": key_lookup,
+            "time_to_lookup_seconds": time.time() - start_time,
+        }
+
+    def _query_bytestring(
+            self,
+            map_index: int,
+            index_in_map: int,
+            collection: bytes,
+            item_size: int
+    ) -> int:
+        map_offset = self._get_map_offset(map_index)
+        return {
+            "map_offset": map_offset,
+            "value": int.from_bytes(
+                collection[
+                    map_offset + index_in_map * item_size:
+                    map_offset + (index_in_map + 1) * item_size
+                ],
+                byteorder="big",
+                signed=False,
+            ),
+            "index_of_value_in_bytestring":
+                map_offset + index_in_map * item_size,
+            "item_size": item_size,
+        }
+
+    def _get_value_at_index(self, map_index: int, index: int) -> int:
+        return self._query_bytestring(
+            map_index,
+            index,
+            self._values,
+            self._value_size,
         )
 
-    def _get_key_at_index(self, index: int) -> int:
-        return int.from_bytes(
-            self._keys[index * self._key_size: (index + 1) * self._key_size],
-            byteorder="big",
-            signed=True,
+    def _get_key_at_index(self, map_index: int, index: int) -> int:
+        return self._query_bytestring(
+            map_index,
+            index,
+            self._keys,
+            self._key_size,
         )
 
-    def _find_index(self, key: int) -> int:
+    def _get_map_offset(self, map_index):
+        # compute map offset based on map index and map sizes
+        return sum(self._map_sizes[:map_index]) * self._key_size
```

Comment:
Should be taking an `item_size` as well like `_query_bytestring`, since the values are 16-bit, rather than 32-bit for the keys.

Reply:
Right! Just fixed it :)